geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java

```diff
@@ -0,0 +1,525 @@
+
+    // Wait for member to get back online.
+    server2.invoke(() -> {
+      cacheRule.createCache();
+      initDataStore(regionShortcut);
+      await().untilAsserted(
+          () -> assertThat(InternalDistributedSystem.getConnectedInstance()).isNotNull());
+    });
+
+    // Let asyncInvocations finish.
+    for (AsyncInvocation<Void> asyncInvocation : asyncInvocationList) {
+      asyncInvocation.await();
+    }
+
+    // Assert Region is empty across VMs.
+    assertRegionIsEmpty(asList(accessor, server1, server2));
+  }
+
+  /**
+   * The test does the following (the clear coordinator is chosen through parameters):
+   * - Sets the {@link MemberKiller} as a {@link CacheWriter} to stop a non-coordinator VM while the
+   * clear is in progress (the member has primary buckets, though, so participates on
+   * the clear operation).
+   * - Launches one thread per VM to continuously execute puts for 30 seconds.
+   * - Clears the Partition Region (at this point the non-coordinator is restarted).
+   * - Asserts that, after the clears are finished, the Region data is consistent across members.
+   */
+  @Test
+  @TestCaseName(TEST_CASE_NAME)
+  @Parameters(method = "vmAndRegionTypes")
+  public void clearWithConcurrentPutsShouldWorkCorrectlyWhenNonCoordinatorMembersAreBounced(
+      TestVM coordinatorVM, RegionShortcut regionShortcut) throws InterruptedException {
+    final int entries = 7000;
+    parametrizedSetup(regionShortcut);
+    registerVMKillerAsCacheWriter(Collections.singletonList(server2));
+
+    // Let all VMs (except the one to kill) continuously execute gets for 60 seconds.
+    List<AsyncInvocation<Void>> asyncInvocationList = Arrays.asList(
+        server1.invokeAsync(() -> executePuts(entries, 60)),
+        accessor.invokeAsync(() -> executePuts(entries, 60)));
+
+    // Clear the region.
+    getVM(coordinatorVM.vmNumber).invoke(() -> cacheRule.getCache().getRegion(REGION_NAME).clear());
+
+    // Wait for member to get back online.
+    server2.invoke(() -> {
+      cacheRule.createCache();
+      initDataStore(regionShortcut);
+      await().untilAsserted(
+          () -> assertThat(InternalDistributedSystem.getConnectedInstance()).isNotNull());
+    });
+
+    // Let asyncInvocations finish.
+    for (AsyncInvocation<Void> asyncInvocation : asyncInvocationList) {
+      asyncInvocation.await();
+    }
+
+    // Assert Region Data is consistent across members.
+    assertRegionDataConsistency(asList(accessor, server1, server2));
+  }
+
+  /**
+   * Shutdowns a member while the clear operation is in progress.
+   * The writer is only installed on the member the test wants to shutdown, doesn't matter whether
+   * it's the clear coordinator or another member holding primary buckets.
+   *
+   * TODO: Review once GEODE-7678 / GEODE-7912 are merged (CacheWriter lifecycle might change).
+   */
+  public static class MemberKiller extends CacheWriterAdapter<String, String> {
```

Comment:
Instead of CacheWriter, you can use DistributionMessageObserver. And can use VM.bounce* to kill vms. No need to wait for CacheWriter messaging implementation.

Reply:
Same as above, I'm still inclined to use the `CacheWriter` for now instead of the `DistributionMessageObserver` hook. If the lifecycle of the listener changes, I'll be more than happy to use another approach :-). If you think this should block the PR from being merged, though, I'll explore using the `DistributionMessageObserver`.

Reply:
I don't think lifecycle of listeners changing, your choice.